text/3642-thread-spawn-hook.md

```diff
@@ -0,0 +1,263 @@
+state even across spawned threads, but a logging/debugging/tracing library could
+keeps track of which thread spawned which thread to provide more useful
+information to the user.
+
+# Public Interface
+
+```rust
+// In std::thread:
+
+/// Registers a function to run for every new thread spawned.
+///
+/// The hook is executed in the parent thread, and returns a function
+/// that will be executed in the new thread.
+///
+/// The hook is called with the `Thread` handle for the new thread.
+///
+/// If the hook returns an `Err`, thread spawning is aborted. In that case, the
+/// function used to spawn the thread (e.g. `std::thread::spawn`) will return
+/// the error returned by the hook.
+///
+/// Hooks can only be added, not removed.
+///
+/// The hooks will run in order, starting with the most recently added.
+///
+/// # Usage
+///
+/// ```
+/// std::thread::add_spawn_hook(|_| {
+///     ..; // This will run in the parent (spawning) thread.
+///     Ok(move || {
+///         ..; // This will run it the child (spawned) thread.
+///     })
+/// });
+/// ```
+///
+/// # Example
+///
+/// ```
+/// thread_local! {
+///     static MY_THREAD_LOCAL: Cell<u32> = Cell::new(0);
+/// }
+///
+/// std::thread::add_spawn_hook(|_| {
+///     // Get the value of X in the spawning thread.
+///     let value = MY_THREAD_LOCAL.get();
+///
+///     Ok(move || {
+///         // Set the value of X in the newly spawned thread.
+///         MY_THREAD_LOCAL.set(value);
+///     })
+/// });
+/// ```
+pub fn add_spawn_hook<F, G>(hook: F)
+where
+    F: 'static + Sync + Fn(&Thread) -> std::io::Result<G>,
+    G: 'static + Send + FnOnce();
+```
+
+# Implementation
+
+The implementation could simply be a static `RwLock` with a `Vec` of
+(boxed/leaked) `dyn Fn`s, or a simple lock free linked list of hooks.
+
+Functions that spawn a thread, such as `std::thread::spawn` will eventually call
+`spawn_unchecked_`, which will call the hooks in the parent thread, after the
+child `Thread` object has been created, but before the child thread has been
+spawned. The resulting `FnOnce` objects are stored and passed on to the child
+thread afterwards, which will execute them one by one before continuing with its
+main function.
+
+# Downsides
+
+- The implementation requires allocation for each hook (to store them in the
+  global list of hooks), and an allocation each time a hook is spawned
+  (to store the resulting closure).
+
+- A library that wants to make use of inheriting thread locals will have to
+  register a global hook, and will need to keep track of whether its hook has
+  already been added (e.g. in a static `AtomicBool`).
+
+- The hooks will not run if threads are spawned through e.g. pthread directly,
+  bypassing the Rust standard library.
+  (However, this is already the case for output capturing in libtest:
+  that does not work across threads when not spawned by libstd.)
+
+# Rationale and alternatives
+
+## Use of `io::Result`.
```

Comment:
A hazard here is that lazy library authors might write an insufficiently contextful error. In particular, they might use the `?` operator to return a real IO error from some kind of per-thread resource creation; such errors are already cryptic, and become more so when separated from their original context.

Therefore, it might make sense to ensure that, if `thread::Builder::spawn()` fails due to a hook, the error it returns is a *wrapped* error whose `to_string()` is something along the lines of "failed to execute thread spawn hook" and whose `Error::source()` is the error the hook returned. (The message could even include the [`Location`](https://doc.rust-lang.org/std/panic/struct.Location.html) of the `set_hook()` call, so that when debugging, one knows _which_ hook to blame.)

If that is done, then the error type can, and perhaps should, be something other than `io::Error`, because there is no advantage to the hook _producing_ an `io::Error` specifically, and making it an IO error might be misleading. (Use `Box<dyn Error + Send + Sync>` if nothing better comes to mind.)